bin/cron/admin_stats

```diff
@@ -4,47 +4,105 @@
 #
 require File.expand_path('../../../pegasus/src/env', __FILE__)
 require src_dir 'database'
-require 'cdo/hip_chat'
 require 'cdo/properties'
 
 DASHBOARD_REPORTING_DB_READONLY = sequel_connect(CDO.dashboard_reporting_db_reader, CDO.dashboard_reporting_db_reader)
-PEGASUS_REPORTING_DB_READONLY = sequel_connect(CDO.pegasus_reporting_db_reader, CDO.pegasus_reporting_db_reader)
+SEVEN_DAYS = 7 * 24 * 60 * 60
+THIRTY_DAYS = 30 * 24 * 60 * 60
+
+# baseline: The set of user objects whose count is used in the denominator of the fraction.
+# comparison: The set of user objects whose count is used in the numerator of the fractions.
+# Returns a hash, mapping demographics (all, gender, locale) to the ratio of users in the
+# comparison and the baseline.
+def build_demographic_breakdowns(baseline, comparison)
+  students = comparison.count.to_f / [1, baseline.count].max.to_f
+  female_students = comparison.where(gender: 'f').count.to_f /
+    [1, baseline.where(gender: 'f').count].max.to_f
+  male_students = comparison.where(gender: 'm').count.to_f /
+    [1, baseline.where(gender: 'm').count].max.to_f
+  us_students = comparison.where(locale: 'en-us').count.to_f /
+    [1, baseline.where(locale: 'en-us').count].max.to_f
+  non_us_students = comparison.exclude(locale: 'en-us').count.to_f /
+    [1, baseline.exclude(locale: 'en-us').count].max.to_f
+  return {
+    students: students,
+    female_students: female_students,
+    male_students: male_students,
+    us_students: us_students,
+    non_us_students: non_us_students,
+  }
+end
 
 def main()
   return unless only_one_running?(__FILE__)
 
-  users = DASHBOARD_REPORTING_DB_READONLY[:users]
-  students = users.where(:user_type => 'student')
-  teachers = users.where(:user_type => 'teacher')
+  # Users, born and unborn.
+  all_users = DASHBOARD_REPORTING_DB_READONLY[:users]
+  all_students = all_users.where(user_type: 'student')
+  all_teachers = all_users.where(user_type: 'teacher')
+  # Born users.
+  users = all_users.exclude(current_sign_in_at: nil)
+  students = users.where(user_type: 'student')
+  teachers = users.where(user_type: 'teacher')
+  # Active students.
+  seven_day_active_students = students.where(current_sign_in_at: (Time.now - SEVEN_DAYS)..Time.now)
+  thirty_day_active_students = students.
+    where(current_sign_in_at: (Time.now - THIRTY_DAYS)..Time.now)
+  # Retention of students...
+  # ...for measuring the percentage of students with accounts at least thirty-seven days and logged
+  #    in at least once after thirty days.
+  thirty_day_retention_historical_baseline = students.
+    where(created_at: Time.new(1970, 1, 1)..(Time.now - (THIRTY_DAYS + SEVEN_DAYS)))
+  thirty_day_retention_historical = thirty_day_retention_historical_baseline.
+    where("current_sign_in_at - created_at < ?", %q{interval '30 days'})
```

Comment:
Maybe I'm missing something but to me this doesn't seem like the query matches the comment?
To me:
* thirty_day_retention_historical_baseline = students who created their accounts more than 37 days ago
* thirty_day_retention_historical = students who created their accounts more than 37 days ago and have not logged in after 30 days

Should the < actually be >? or is there a 1- that I missed somewhere?